[ONNX] Support converting fx graph with symbolic shape to ONNX

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97494
* __->__ #96350

~~Need https://github.com/microsoft/onnx-script/pull/484~~

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch between symbolic tracing (dynamic sizes) and fake mode (static). Set it to default True, as most of our tests are happy with sumbolic tracing. Except GPT2 stays with fake mode with error: https://github.com/microsoft/onnx-script/issues/523
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter. This can be removed once tests are integrated with https://github.com/pytorch/pytorch/pull/96479
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added to support SymFloat/SymInt. (FIXME: https://github.com/pytorch/pytorch/issues/97201). sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.
4. Fully deprecated FakeTensorProp as make_fx() should provide all node meta info.
5. Put complicated fx.Node related ArgumentType into _type_utils.py

titaiwangms committed:
[ONNX] Support dynamic axes

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,382 @@
+            DynamicAdd(), (x, y), additional_test_inputs=[(input_x, input_y)]
+        )
+
+    @unittest.skip("flaky test: sometimes SegFault on onnx_model.SerializeToString()")
```
This is a weird case.

titaiwangms replied:
https://github.com/microsoft/onnx-script/issues/523

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -316,6 +351,25 @@ def _export_fx_node_to_onnxscript(
+        elif (
+            isinstance(node.target, types.BuiltinFunctionType)
+            and node.name.split("_")[0]
+            in function_dispatcher._BUILTIN_OVERLOAD_TO_EXPORTER_KEY_TABLE
```
Curious why this is a special case?

justinchuby commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -180,13 +180,23 @@ def _create_op_overload_to_exporter_key_table() -> Dict[
 # _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE[torch.add.Tensor] is "aten::add".
 _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE = _create_op_overload_to_exporter_key_table()
 
+_BUILTIN_OVERLOAD_TO_EXPORTER_KEY_TABLE = {
```
I would make public if it is used by other modules.

titaiwangms replied:
As well as _ATENLIB_XXX?

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -38,13 +37,21 @@ def _export(
     # decomposed_module.graph.nodes[i].meta["val"]
     decomposed_module = passes.shape_inference_with_fake_tensor(
-        decomposed_module, *args
+        decomposed_module, *inputs
```
@wschin LMK if this makes sense

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -276,11 +352,19 @@ def _export_fx_node_to_onnxscript(
+            output = onnxscript_graph.add_input(
+                input_name=None,
+            )
+        else:
```
@justinchuby A cost of not having Tensor as argument...

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

wschin commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -121,6 +123,11 @@ def export_after_normalizing_args_and_kwargs(
+        op_level_debug: Whether to export the model with op level debug information
+            with onnxruntime evaluator.
+        dynamic_axes: Whether to export the model with dynamic axes. This would set
+            the shape of input and nodes all to dynamic by following symbolic fx graph.
```
Terms needs to be as explicit as possible. "following symbolic fx graph" is not a well known term.
```suggestion
            the shape of input and nodes all to dynamic by calling `proxy_tensor.make_fx` with `tracing_mode="symbolic"`.
```

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -180,13 +180,23 @@ def _create_op_overload_to_exporter_key_table() -> Dict[
 _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE = _create_op_overload_to_exporter_key_table()
 
+_BUILTIN_OVERLOAD_TO_EXPORTER_KEY_TABLE = {
+    "mul": "aten::mul",
```
Don't do this. It will map all `+`, `-`, `*`, and `/` to `aten::*`. Let's do
```
_BUILTIN_OVERLOAD_TO_EXPORTER_KEY_TABLE = {
    "mul": "__builtin_mul__",
    "add": "__builtin_add__",
    "sub": "__builtin_sub__",
    "truediv": "__builtin_truediv__",
}
```

Then, create the corresponding functions in dispatch table such as
```
def __builtin_mul__(self, other):
    if x is not Tensor or y is not Tensor:
      raise
    ops.core.aten_add(x, y)
    
```

Basically, I want to check the types before calling those built's exporters.

wschin replied:
Per our discussion, we can replace this dictionary with
```
# This means all SymInt's are mapped to `torch.Tensor` before mapping ONNX Script.
torch.SymInt.__add__: "aten::add"
...
```

titaiwangms replied:
I use operator.add instead. LMK if it's good.

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +133,33 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+    if isinstance(onnx_tensor, (tuple, list)) and any(
+        isinstance(ele, torch.fx.Node) for ele in onnx_tensor
+    ):
+        # This intends to solve dynamic axes. for example, if the size input of op.Expand
```
```suggestion
        # This intends to handle dynamic axes. for example, if the size input of op.Expand
```

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +133,33 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+        isinstance(ele, torch.fx.Node) for ele in onnx_tensor
+    ):
+        # This intends to solve dynamic axes. for example, if the size input of op.Expand
+        # is dynamic, each of dim would be originated from other tensors.
```
```suggestion
        # is dynamic, each of dim would be variables (e.g, sym variable in Pytorch FX graph. Note that sym variable is mapped to tensors in ONNX Script world) calculated by other operators.
```

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +133,33 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+        #    inputs: [2, 4, TensorA, 1, TensorB]
+        #    outputs: Tensor([Tensor(2), Tensor(4), TensorA, Tensor(1), TensorB])
+
+        # onnx-script auto wrap constants with op.Constants,
```
Humm, I don't see the key explanation in the doc. Please say `shapes are mapped to tensors in ONNX graph (type: TorchScriptGraph?), so list of sym_ints is concatenated to a tensor before calling ONNX op.`

titaiwangms replied:
I was thinking this is not only applied to SymInt, it's for all List[fx.Node] scenario. If I update the comment to specify SymInt, I think we need to add another elif branch for those list of fx.Node without SymInt value?

wschin replied:
I don't have a good idea to handle `all` cases but just having a specific branch for sym_* sounds good enough.

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +133,33 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
-    elif isinstance(onnx_tensor, torch.dtype):
+        return fx_name_to_onnxscipt_value[onnx_tensor.name]
+    if isinstance(onnx_tensor, (tuple, list)) and any(
+        isinstance(ele, torch.fx.Node) for ele in onnx_tensor
```
Are we sure we want to concatenate all `lists of fx.Node`?

justinchuby replied:
yeah we need to be careful about this

wschin replied:
What if there is a `[Tensor<bool>, Tensor<float>]` (e.g., dropout_backward's inputs)? There can be other forward ops with hybrid element types.

titaiwangms replied:
Make sense

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -164,12 +195,16 @@ def _fill_tensor_meta(
     """Fill the meta information of onnxscript_values with that from the fx FakeTensor."""
 
+    # aten::sym_size output is a int, not a tensor, which stands
+    # for the size of one dim. We treat it as 1D tensor.
```
1-D tensor or 0-D tensor? Why not 0-tensor?

titaiwangms replied:
Done

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -228,17 +272,49 @@ def _wrap_fx_args_as_onnxscript_args(
-                        torch_args.append(
-                            torch.randn_like(meta_value, dtype=torch.float)
+                fake_tensor = arg.meta["val"]
+                if isinstance(fake_tensor, list):
```
Too hacky...and it will go wrong when there are shape-realted ops. Let's not to verify when dynamic axis is on.

titaiwangms replied:
This is actually for op level debugging, I will split this part out.

wschin reviewed (changes requested):
I left some comments about
1. Document (need more explanation on the overall design. As you may see in my comments, there are some dangerous code. Without proper document, it's hard for other reviewers to capture them)
2. Please proof read the whole code with document.
3. Some places look unsafe. We need to improve them.

Please feel free to ping me to discuss. I might not check github email regularly. Thank you.

justinchuby commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -13,7 +14,10 @@
 
 import onnxscript  # type: ignore[import]
-from onnxscript import evaluator  # type: ignore[import]
+from onnxscript import (  # type: ignore[import]  # type: ignore[import]
```
duplicated?

titaiwangms replied:
Done

justinchuby commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +133,33 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+    if isinstance(onnx_tensor, torch.dtype):
         onnx_tensor = int(_type_utils.JitScalarType.from_dtype(onnx_tensor).onnx_type())
 
+    # all other cases, we does nothing.
```
nit: grammar

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -204,6 +298,13 @@ def _wrap_fx_args_as_onnxscript_args(
+    elif isinstance(node.target, torch._ops.OpOverloadPacket):
+        # aten::sym_size is the only OverloadPacket that we support.
+        # schema: aten::sym_size(Tensor self, int dim) -> Tensor
+        assert str(node.target) == "aten.sym_size"
```
Add an error message, or change to raise exception with error message.

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -204,6 +298,13 @@ def _wrap_fx_args_as_onnxscript_args(
+    elif isinstance(node.target, torch._ops.OpOverloadPacket):
+        # aten::sym_size is the only OverloadPacket that we support.
+        # schema: aten::sym_size(Tensor self, int dim) -> Tensor
+        assert str(node.target) == "aten.sym_size"
```
Is this "elif" branch mainly for the assertion check?

titaiwangms replied:
No. As OpOverloadPacket doesn't have schema, it would do nothing if we don't handle it specifically.

BowenBao replied:
Discussed offline, root cause #97201.

titaiwangms replied:
Now it's manually pointing to torch.ops.aten.sym_int.int

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -178,23 +217,78 @@ def _fill_tensor_meta(
+def _wrap_fx_args_as_torch_args(
+    complete_args: list, complete_kwargs: dict
+) -> Tuple[tuple, dict]:
+    # prepare torch format args and kwargs for op-level validation
```
nit: Capitalize first letter of a sentance in a comment for readability and consistency. I would make it a docstring in this case

 ```suggestion
    """Prepare torch format args and kwargs for op-level validation."""
```

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -37,15 +36,17 @@ def _export(
-
+    # TODO(titaiwang): in fake mode we have seen fx.nodes without node.meta["val"]
+    # Do we need this in symbolic mode?
+    if not export_options.dynamic_axes:
```
@BowenBao @wschin https://github.com/pytorch/pytorch/pull/96879#issuecomment-1477181945

titaiwangms replied:
So only one of the `fake` mode test would lack of node.meta["val"]. I guess if all nodes should be expected to have node.meta["val"], we should look into model instead? (give up this pass)

titaiwangms replied:
I killed the pass

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +130,40 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+        # is dynamic, each dimension would be variable (i.e., sym variable in Pytorch
+        # FX graph. Note that sym variable is mapped to tensor in ONNX Script world)
+        # calculated by other operators.
+        sequence_elements = []
```
type annotation

titaiwangms replied:
DOne

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +130,40 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+
+        # For example:
+        #    inputs: [2, 4, TensorA, 1, TensorB]
+        #    outputs: Tensor([Tensor(2), Tensor(4), TensorA, Tensor(1), TensorB])
```
```suggestion
        #    outputs: onnx::Concat([Tensor(2), Tensor(4), TensorA, Tensor(1), TensorB])
```
I wonder if this makes more sense.

titaiwangms replied:
I was kind of referring Tensor in this case as onnx tensor.

BowenBao replied:
Is this more accurate? `Tensor([2, 4, TensorA.item(), 1, TensorB.item()])`. I thought about commenting this, but then I wonder if you used `onnx::Concat` to achieve the onnx tensor.

The original line shows nested tensor, which is not a legal representation.

titaiwangms replied:
How about this:

```python
# For example:
#    inputs: [2, 4, fx.Node(SymIntA), 1, fx.Node(SymIntB)]
#    outputs: op.Concat([op.Constant(2), op.Constant(4), TorchScriptTensor(A), op.Constant(1), TorchScriptTensor(B)])
```

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +130,40 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
+        #    inputs: [2, 4, TensorA, 1, TensorB]
+        #    outputs: Tensor([Tensor(2), Tensor(4), TensorA, Tensor(1), TensorB])
+
+        # onnx-script auto wrap constants with op.Constants,
```
```suggestion
        # onnx-script auto wrap python numbers with op.Constants,
```
Is this more accurate?

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -178,23 +217,78 @@ def _fill_tensor_meta(
+    torch_args = []
+    for arg in complete_args:
+        if isinstance(arg, torch.fx.Node):
+            # Create a concreate test tensor based on the fake tensor
```
```suggestion
            # Create a concrete test tensor based on the fake tensor
```
This comment looks repeated to line 237.

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -178,23 +217,78 @@ def _fill_tensor_meta(
+                        torch_args.append(torch.randn(meta_value.shape, dtype=dtype))
+                elif isinstance(fake_tensor, torch.Tensor):
+                    # could be varying size
+                    # use 2 to replace dynamic shape
```
😱

BowenBao replied:
https://github.com/microsoft/onnx-script/issues/393#issuecomment-1478289501

justinchuby replied:
Lolll

> 😱

titaiwangms replied:
Yup, this is a mess. I intend to address this in another PR. Currently, this is under op_level_debug. Use FakeTensorProp on symbolic graph might be the solution!

wschin replied:
No, `2` is one thing should never appears in the codebase. As suggested in my very early comment, this should be an optional feature and disabled when dynamic shape is enabled. This is reasonable because when there is no real shape, we don't know what inputs to execute on.

titaiwangms replied:
It's done, please see the options.py

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -178,23 +217,78 @@ def _fill_tensor_meta(
 
+def _wrap_fx_args_as_torch_args(
+    complete_args: list, complete_kwargs: dict
+) -> Tuple[tuple, dict]:
```
nit: similar to above comment on type annotations. Would greatly help if annotated with more details. e.g. `Tuple[Tuple[torch.Tensor, ...], Dict[str, torch.Tensor]]`

titaiwangms replied:
I introduced Annotation from torch/fx/node.py, and put it into _type_utils.py. LMK if it's good enough.

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -276,11 +348,19 @@ def _export_fx_node_to_onnxscript(
+        # NOTE: add_input() intends to create nodes with shape/type
+        fake_tensor = node.meta["val"]
+        if fake_tensor is None:
+            output = onnxscript_graph.add_input(
```
It's from the original code. Now I look at it, we are not supposed to have `None` in any of node.meta["val"]. Do you think we should error here?

BowenBao replied:
Weren't there another comment saying there is a test that could lack `meta["val"]`? If this is not a valid onnx input, please do error.

titaiwangms replied:
It seems like we have the test: test_large_scale_exporter_with_tiny_gpt2 with intentionally added None as input, which is making edge case here, and need FakeTensorProp.

If I do like:
```python
fake_tensor = node.meta.get("val", None)
```

We can totally kill shape_inference.py pass. It seems only fill in the None for us in our tests...

titaiwangms replied:
shape_inference.py is deprecated, but not deleted. I will try to utilize it in op_level_debug refactor.

titaiwangms committed:
Update on "[ONNX] Support dynamic axes"


Need https://github.com/microsoft/onnx-script/pull/484

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch to avoid affecting what we have on static export.
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter.
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added.
4. sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -125,10 +130,40 @@ def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_valu
-    elif isinstance(onnx_tensor, torch.dtype):
+        return fx_name_to_onnxscipt_value[onnx_tensor.name]
+    if isinstance(onnx_tensor, (tuple, list)) and any(
+        isinstance(ele, torch.fx.Node) and isinstance(ele.meta["val"], torch.SymInt)
```
change it to node

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -383,15 +511,13 @@ def _export_fx_node_to_onnxscript(
-        input_ = onnxscript_graph.add_input(
-            input_name=node.name, input_value=current_attr
-        )
+        input_ = onnxscript_graph.add_initializer(node.name, current_attr)
```
add_initializer API updated.

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -305,17 +406,37 @@ def _export_fx_node_to_onnxscript(
             fx_name_to_onnxscipt_value[node.name] = output
             return
-
-        if node.target == operator.getitem:
```
Combined into _BUILTIN_TO_EXPORTER_KEY_TABLE

titaiwangms replied:
Reverted

wschin commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+@_beartype.beartype
+# TODO(titaiwang): bound.args makes pytorch_inputs hard to annotate
+# maybe annotate it when the exporter API is launched
+def _run_ort(onnx_model: _ONNXModelType, pytorch_inputs: Any) -> _OutputsType:
```
This type should be an iterable object per code right below.
```suggestion
def _run_ort(onnx_model: _ONNXModelType, pytorch_inputs: Union[Tuple[...], List[...]]) -> _OutputsType:
```

titaiwangms replied:
Added annotations

wschin commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+
+
+@_beartype.beartype
+def _run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
```
Missing document. I don't know what are `additional_test_inputs` and `additional_test_kwargs` for.

titaiwangms replied:
Done

wschin commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+                "Failed to clone model. Model state might be mutated during verification."
+            )
+            return model
+
```
I am not sure what's the standard to use `beartype`. Is it a requirement for ALL functions? If yes, maybe annotate `compare_pytorch_onnx_with_ort` as well.

titaiwangms replied:
Done

wschin commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+    atol: float = 1e-7,
+    opset_version: int = 18,
+    additional_test_inputs: Optional[Sequence[_InputArgsType]] = None,
+    additional_test_kwargs: Optional[Sequence[_InputKwargsType]] = None,
```
Oops. If this function is documented, you will notice `additional_test_kwargs` is not used below and can be deleted.

titaiwangms replied:
Done

wschin commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+
+        x = torch.randn(2, 3)
+        y = torch.randn(2, 3)
+        input_x = torch.randn(3, 4)
```
```suggestion
        another_x = torch.randn(3, 4)
```

wschin replied:
It's kind of confusing seeing two input `x`..

titaiwangms replied:
Done

wschin commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -227,6 +233,7 @@ def _test_large_scale_exporter(
         create_model: Callable,
         create_args: Callable,
         create_pytorch_only_kwargs: Callable,
+        dynamic_axes: bool = True,
```
```suggestion
        enable_dynamic_shape: bool = True,
```

justinchuby replied:
I would actually vote for dynamic_axes. I think this way of naming is more conventional in Python where we don’t say “enable”, “use”, “is” etc. Aka options don’t quite start with verbs.

justinchuby replied:
As an argument, it can sometimes be Union[bool, Dict[int, str]] for example. It will just be a matter of not choosing a conflicting name in the function body. 

For example, we can name the argument dynamic_shape so there’s no conflict

justinchuby replied:
Now dynamic shape can be less accurate than dynamic axes because dynamic shape also implies the rank is dynamic?

wschin replied:
It's ok to use `enable_dynamic_axes`. However, as mentioned above, code with dynamic shape behavior can happen 100 times more than the use of this flag. Which one do you think should take the better name? Imagining that if I author 1000 custom ops (not rare because aten has more ops) for dynamic shape, do you want the 1000 custom ops to use `local_dynamic_axes` or just `dynamic_axies` in theyr core logics?

titaiwangms replied:
So I checked the old API, and we did use verb with bool keyword arguments, I will align with that format. However, I will put a NOTE for us to discuss this again when the real exporter API PR is out.

wschin replied:
ChatGPT likes `enable_dynamic_axes`.

By ChatGPT:

Both options can be used to name a boolean variable that turns on/off a feature in Python. However, enable_dynamic_axes is generally a more commonly used naming convention in Python.

The use of the word "enable" in the variable name makes it clear that the variable controls whether or not a feature is enabled. Additionally, by using the prefix "enable_", it follows the convention of prefixing boolean variable names with verbs that indicate the action the variable controls. This is a common naming convention in Python, making code more readable and easier to understand for other developers.

On the other hand, dynamic_axes does not clearly indicate that it is a boolean variable and what it controls. Therefore, it might cause confusion to other developers who are not familiar with the code.

Therefore, I would recommend using enable_dynamic_axes as the variable name.

justinchuby replied:
> It's ok to use `enable_dynamic_axes`. However, as mentioned above, code with dynamic shape behavior can happen 100 times more than the use of this flag. Which one do you think should take the better name? Imagining that if I author 1000 custom ops (not rare because aten has more ops) for dynamic shape, do you want the 1000 custom ops to use `local_dynamic_axes` or just `dynamic_axies` in theyr core logics?

I would like to point out two things from this scenario: 1) is that we don’t expect this argument to suddenly appear in 1000 ops, if at all. It is an exporter toggle and the ops are expected to support, on or off. 2) the naming concerns mostly the public interfaces and that’s usually a few levels higher than where the logic actually is. We should still be able to strike a balance and not resort to something like `local_dynamic_axes`

justinchuby replied:
I also overlooked that this is an internal function, in which case I don’t have a strong opinion

titaiwangms replied:
Right this not yet a public API. I am resolving this conversation.

wschin commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -349,7 +356,11 @@ def create_pytorch_only_extra_kwargs():
+            create_model,
+            create_args,
+            create_pytorch_only_extra_kwargs,
+            dynamic_axes=True,
```
Let's test both of `dynamic_axes=True` and `dynamic_axes=False`.

titaiwangms replied:
I thought symbolic mode is the top choice of exporter? I see dynamic_axes=False(fake mode) as a back plan. In other words, why do we need to main fake mode if we know symbolic mode is working?

wschin replied:
If you don't test backup plan, how do you know backup plan will work?

titaiwangms replied:
Well, I was thinking we don't need other modes anymore if we have symbolic. But in these two large scale cases, they need to export with FakeTensorMode, which it seems to disallow us use symbolic in later dynamo/make_fx. So either way, this should be False.

BowenBao commented:
I updated the title to `fx graph with symbolic shape`. `symbolic tracing` was an overloaded term, the legacy fx tracer was called `symbolic trace`.

wschin commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -178,14 +180,24 @@ def _create_op_overload_to_exporter_key_table() -> (
 # Dictionary that maps torch.ops.aten.* to exporter look up key; e.g.,
 # _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE[torch.add.Tensor] is "aten::add".
 _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE = _create_op_overload_to_exporter_key_table()
+_BUILTIN_TO_EXPORTER_KEY_TABLE = {
```
```suggestion
_SYMINT_SYMFLOAT_BUILTIN_TO_EXPORTER_KEY_TABLE = {
```

titaiwangms replied:
Reverted

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -305,17 +406,37 @@ def _export_fx_node_to_onnxscript(
+        ):
+            # This function expects built-in functions to have name like mul, mul_1, ...
+            # symbolic fx.graph contains built-in functions to calculate python values.
+            exporter_key = function_dispatcher._BUILTIN_TO_EXPORTER_KEY_TABLE[
```
Can you assert if all inputs (in `node.args`) are tensor, int, float, and symint? I don't want to dispatch a random object such as `symfloat` to `aten::add` at this stage.

titaiwangms replied:
Why not allow SymFloat?

titaiwangms replied:
I have seen SymFloat when I tried to get symbolic mode of GPT2. It was used with __pow__. so I added it (GPT2 has some other exporting problem though). But if you have a strong reason of disallowing SymFloat for now, please LMK, and I will remove them.

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -305,17 +406,37 @@ def _export_fx_node_to_onnxscript(
+            # schema: aten::sym_size(Tensor self, int dim) -> Tensor
+            if node.target != torch.ops.aten.sym_size:
+                raise ValueError(
+                    f"Unsupported OverloadPacket: {node.target}, aten.sym_size is the only allowed OverloadPacket!"
```
Great! As restrictive as possible in dev facing code (as convenient as possible in user facing code)!

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -305,17 +406,37 @@ def _export_fx_node_to_onnxscript(
+            # We manually assigned overload for aten::sym_size.
+            node.target = torch.ops.aten.sym_size.int
+            exporter_key = function_dispatcher._OP_OVERLOAD_TO_EXPORTER_KEY_TABLE[
+                node.target  # type: ignore[index]
```
```suggestion
                torch.ops.aten.sym_size.int  # type: ignore[index]
```
Would this avoid mutation to the node? I think we should avoid implicit graph mutation whenever possible in a graph transformation pass (fx_to_onnx is kind of such pass).

titaiwangms replied:
Done. 

But this is needed in `_wrap_fx_args_as_onnxscript_args`:

```python
    # TODO(titaiwang): aten::sym_size has overload, but fx graph is using
    # overloadpacket for some reasons.
    # https://github.com/pytorch/pytorch/issues/97201
    # We manually assigned overload for aten::sym_size.
    if hasattr(node.target, "_schema"):
        node_schema = node.target._schema
    else:
        node_schema = torch.ops.aten.sym_size.int._schema
```

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -111,7 +112,11 @@ def _location_from_fx_stack_trace(
 
-def _retrieve_or_adapt_input_to_graph_set(fx_node_arg, fx_name_to_onnxscipt_value):
+def _retrieve_or_adapt_input_to_graph_set(
+    fx_node_arg,
```
Done

wschin commented on torch/onnx/_internal/fx/options.py:
```diff
@@ -33,3 +40,10 @@ def update(self, **kwargs):
+        # op_level_debug would be forced to False if dynamic_axes is True
+        # https://github.com/microsoft/onnx-script/issues/393
+        if self.dynamic_axes and self.op_level_debug:
+            self.op_level_debug = False
```
No. Let's assert with clear instruction. Implicitly changining user's direct input is not a good idea.
```
if op_level_debug:
  assert enable_dynamic_shape == False, "op_level_debug and enable_dynamic_shape are mutually execusive. Please set only one of them to be True"
  ```

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -178,14 +180,24 @@ def _create_op_overload_to_exporter_key_table() -> (
+    # TODO(titaiwang): aten::sym_size has overload, but fx graph is using
+    # overloadpacket for some reasons.
+    # https://github.com/pytorch/pytorch/issues/97201
+    table[torch.ops.aten.sym_size.int] = "aten::sym_size"
```
We went over this offline. Did you find that this is still required? i.e. `# NOTE: Below are not in torch.ops.aten/torch.ops.prim` this is true for `torch.ops.aten.sym_size.int`?

To be precise, the below 3 lines of comment is unrelated to this particular line, and should be put in `fx_to_onnxscript.py` (which you already did).
```
    # TODO(titaiwang): aten::sym_size has overload, but fx graph is using
    # overloadpacket for some reasons.
    # https://github.com/pytorch/pytorch/issues/97201
```

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Support converting fx graph with symbolic shape to ONNX"


~~Need https://github.com/microsoft/onnx-script/pull/484~~

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch between symbolic tracing (dynamic sizes) and fake mode (static). Set it to default True, as most of our tests are happy with sumbolic tracing. Except GPT2 stays with fake mode with error: https://github.com/microsoft/onnx-script/issues/523
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter. This can be removed once tests are integrated with https://github.com/pytorch/pytorch/pull/96479
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added to support SymFloat/SymInt. (FIXME: https://github.com/pytorch/pytorch/issues/97201). sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.
4. Fully deprecated FakeTensorProp as make_fx() should provide all node meta info.
5. Put complicated fx.Node related ArgumentType into _type_utils.py

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+        input_x = torch.randn(3, 4)
+        input_y = torch.randn(3, 4)
+
+        _run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
```
Done

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -347,7 +468,14 @@ def _export_fx_node_to_onnxscript(
             output
         )
         if options.op_level_debug:
```
Done

titaiwangms commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -0,0 +1,384 @@
+_ONNXModelType = Union["onnx.ModelProto", bytes, str, io.BytesIO]
+_InputArgsType = Union[torch.Tensor, Tuple[Any, ...]]
+_InputKwargsType = Mapping[str, Any]
+_OutputsType = Union[Sequence[_NumericType], Sequence]
```
Removed

titaiwangms commented:
@wschin PTAL

wschin commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -305,17 +408,45 @@ def _export_fx_node_to_onnxscript(
+            in function_dispatcher._SYMINT_SYMFLOAT_BUILTIN_TO_EXPORTER_KEY_TABLE
+        ):
+            if not isinstance(
+                node.meta["val"], (int, float, torch.SymInt, torch.SymFloat)
```
nit
```
if not all(node_arg.meta["val"] in (int, float, torch.SymInt, torch.SymFloat):
  raise ...
```
Dispatch should only depend on input types.

titaiwangms committed:
Update on "[ONNX] Support converting fx graph with symbolic shape to ONNX"


~~Need https://github.com/microsoft/onnx-script/pull/484~~

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch between symbolic tracing (dynamic sizes) and fake mode (static). Set it to default True, as most of our tests are happy with sumbolic tracing. Except GPT2 stays with fake mode with error: https://github.com/microsoft/onnx-script/issues/523
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter. This can be removed once tests are integrated with https://github.com/pytorch/pytorch/pull/96479
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added to support SymFloat/SymInt. (FIXME: https://github.com/pytorch/pytorch/issues/97201). sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.
4. Fully deprecated FakeTensorProp as make_fx() should provide all node meta info.
5. Put complicated fx.Node related ArgumentType into _type_utils.py

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support converting fx graph with symbolic shape to ONNX"


~~Need https://github.com/microsoft/onnx-script/pull/484~~

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch between symbolic tracing (dynamic sizes) and fake mode (static). Set it to default True, as most of our tests are happy with sumbolic tracing. Except GPT2 stays with fake mode with error: https://github.com/microsoft/onnx-script/issues/523
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter. This can be removed once tests are integrated with https://github.com/pytorch/pytorch/pull/96479
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added to support SymFloat/SymInt. (FIXME: https://github.com/pytorch/pytorch/issues/97201). sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.
4. Fully deprecated FakeTensorProp as make_fx() should provide all node meta info.
5. Put complicated fx.Node related ArgumentType into _type_utils.py

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support converting fx graph with symbolic shape to ONNX"


~~Need https://github.com/microsoft/onnx-script/pull/484~~

Support dynamic export on fx-ONNX exporter. Essentially, we set inputs size and nodes all dynamic in torchscript, and leverage on `aten::sym_size` to catch dynamic size between each Op.

1. Add `dynamic_axes` switch between symbolic tracing (dynamic sizes) and fake mode (static). Set it to default True, as most of our tests are happy with sumbolic tracing. Except GPT2 stays with fake mode with error: https://github.com/microsoft/onnx-script/issues/523
2. Add test_fx_dynamic_onnruntime.py to test on some addhoc we have from old exporter. This can be removed once tests are integrated with https://github.com/pytorch/pytorch/pull/96479
3. Since `aten::sym_size` are operated with built-in function, a built-in function mapping is added to support SymFloat/SymInt. (FIXME: https://github.com/pytorch/pytorch/issues/97201). sym_size output value is also fx.Node, and can be found in `fx_name_to_onnxscipt_value`, so it's operation stays the same as other ONNX ops in ONNX graph.
4. Fully deprecated FakeTensorProp as make_fx() should provide all node meta info.
5. Put complicated fx.Node related ArgumentType into _type_utils.py

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge -g

titaiwangms commented:
@pytorchbot merge

titaiwangms commented:
@pytorchbot merge
